pkg/apis/flows/v1alpha1/flow_validation_test.go

```diff
@@ -0,0 +1,185 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"github.com/knative/pkg/apis"
+	"k8s.io/api/core/v1"
+	"testing"
+)
+
+func TestFlowSpecValidation(t *testing.T) {
+	tests := []struct {
+		name string
+		f    *FlowSpec
+		want *apis.FieldError
+	}{{
+		name: "valid with URI",
+		f: &FlowSpec{
+			Action: FlowAction{
+				TargetURI: stringPtr("http://example.com/action"),
+			},
+			Trigger: EventTrigger{
+				EventType: "foo",
+				Resource:  "bar",
+			},
+		},
+	}, {
+		name: "valid with Target",
+		f: &FlowSpec{
+			Action: FlowAction{
+				Target: &v1.ObjectReference{
+					Name: "foo",
+				},
+			},
+			Trigger: EventTrigger{
+				EventType: "foo",
+				Resource:  "bar",
+			},
+		},
+	}, {
+		name: "invalid with URI",
+		f: &FlowSpec{
+			Action: FlowAction{
+				TargetURI: stringPtr("http//example.com/action"),
+			},
+			Trigger: EventTrigger{
+				EventType: "foo",
+				Resource:  "bar",
+			},
+		},
+		want: &apis.FieldError{
+			Message: `invalid value "http//example.com/action"`,
+			Paths:   []string{"action.targetURI"},
```

Comment:
I am still catching up with the changes for validation code while I was gone. I like what's being done with them, I'm just little confused on what the eventual returned to the user and how the paths get stitched together. But that's not really part of this PR, just need to go read some more code :)

Reply:
The idea is that the field context is prefixed as you come back up the call stack, so you can traverse a object graph and call validate on sub-objects. If those objects throw errors, you add the context only you know about in the context the object is used. IE:

```
err = obj.AnotherObj.Validate()
if err != nil {
  return err.ViaField("AnotherObj")
}
```

AnotherObj's type was able to have a validate function that is only aware of the objects requirements. Only obj knew the context in which the user knows how they used AnotherObj.

Check out `knative/pkg/apis`